zallet/src/components/sync/steps.rs

```diff
@@ -0,0 +1,345 @@
+                        "Invalid subtree root: {}",
+                        e
+                    )),
+                )
+            })?;
+            Ok(CommitmentTreeRoot::from_parts(
+                BlockHeight::from_u32(subtree.end_height.0),
+                sapling::Node::from_bytes(root_hash).unwrap(),
+            ))
+        })
+        .collect::<Result<Vec<_>, SyncError>>()?;
+
+    info!("Sapling tree has {} subtrees", sapling_roots.len());
+    db_data.put_sapling_subtree_roots(0, &sapling_roots)?;
+
+    let orchard_roots = chain
+        .z_get_subtrees_by_index("orchard".into(), NoteCommitmentSubtreeIndex(0), None)
+        .await?
+        .subtrees
+        .into_iter()
+        .map(|subtree| {
+            let mut root_hash = [0; 32];
+            hex::decode_to_slice(&subtree.root, &mut root_hash).map_err(|e| {
+                FetchServiceError::JsonRpcConnectorError(
+                    JsonRpSeeConnectorError::JsonRpSeeClientError(format!(
+                        "Invalid subtree root: {}",
+                        e
+                    )),
+                )
+            })?;
+            Ok(CommitmentTreeRoot::from_parts(
+                BlockHeight::from_u32(subtree.end_height.0),
+                MerkleHashOrchard::from_bytes(&root_hash).unwrap(),
+            ))
+        })
+        .collect::<Result<Vec<_>, SyncError>>()?;
+
+    info!("Orchard tree has {} subtrees", orchard_roots.len());
+    db_data.put_orchard_subtree_roots(0, &orchard_roots)?;
+
+    Ok(())
+}
+
+pub(super) async fn get_chain_tip(chain: &FetchServiceSubscriber) -> Result<ChainBlock, SyncError> {
+    ChainBlock::tip(chain).await
+}
+
+pub(super) async fn find_fork(
+    chain: &FetchServiceSubscriber,
+    mut prev_tip: ChainBlock,
+    mut current_tip: ChainBlock,
+) -> Result<ChainBlock, SyncError> {
+    // Roll tips backwards until they have the same height.
+    while prev_tip.height > current_tip.height {
+        prev_tip =
+            ChainBlock::resolve(chain, prev_tip.prev_hash.expect("present by invariant")).await?;
+    }
+    while current_tip.height > prev_tip.height {
+        current_tip =
+            ChainBlock::resolve(chain, current_tip.prev_hash.expect("present by invariant"))
+                .await?;
+    }
+
+    // Roll tips backwards until they are the same block.
+    while prev_tip != current_tip {
+        // We are fetching blocks from the same data source, and we assume here that the
+        // data source has one single block with no parent (the genesis block). Therefore
+        // if the blocks are not currently equal, they cannot be the genesis block, and so
+        // their parent hashes exist (per the `ChainBlock` invariant).
+        prev_tip =
+            ChainBlock::resolve(chain, prev_tip.prev_hash.expect("present by invariant")).await?;
+        current_tip =
+            ChainBlock::resolve(chain, current_tip.prev_hash.expect("present by invariant"))
+                .await?;
```

Comment:
These also should not be panics.

Reply:
These panics are fine, because if they fail it is a programming errors (namely a `ChainBlock` being constructed in violation of its documented invariant that `prev_hash` is `None` if and only if `height` is `consensus::H0`).